Improve JSX Formatting

suchipi opened the pull request:
* Arrow Function Expressions returning JSX will now add parens when the JSX breaks
* Conditional expressions within (or containing) JSX are formatted in a more natural way (for JSX), especially when chained
* JSX in logical expressions (|| or &&) is ~~always wrapped in parens~~ wrapped in parens when the JSX breaks
* `null` in conditional expressions within (or containing) JSX is never wrapped in parens

Fixes #2208

TODO:
- [x] Fix JSX-mode ternaries always breaking
- [x] Add comments to tests and make sure I have good coverage

suchipi commented:
Test failure appears to be unrelated; it's also broken in f490dba60926f2546c7a29ef33c7ec9322968eaa

azz commented:
Yeah we have a flaky test, haven't got a chance to figure out the cause. #2361

I restarted the build for you.

ipanasenko commented on tests/last_argument_expansion/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -109,19 +109,21 @@ true
+  test({
+    a: 1
+  })
+) : (
```
In my opinion, current format looks better

wmertens replied:
I used to like this format because it works well with editors and manual formatting, but now that prettier handles formatting I ended up preferring the prettier format. It is "lighter" and still unambiguous.

zwily replied:
See #2208 for the functional argument for changing it.

TrySound commented on tests/jsx/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -228,6 +311,62 @@ exports[`hug.js 1`] = `
+</div>;
+
+<div>
+  {a && <span>
```
I think before multiline or too long jsx should be linebreak like
```js
(
  <div>
    {props.header &&
      <header>
      </header>
    }
  </div>
);
```

suchipi replied:
You commented on the unformatted code, so I'm a bit confused.

Current behavior in this PR is to always wrap parens (and add line breaks) around JSX in `&&`/`||`, whether it's breaking or not.

TrySound replied:
Me too. Isn't it snapshot with result?

suchipi replied:
The snapshot contains both the unformatted and formatted code, with a line of `~` characters in-between them.

The formatted code is on [line 354](https://github.com/prettier/prettier/pull/2398/files/e5254e78f13518f488205789af560af12a7fcbb0#diff-4591a84e8b3362fa01bb47a987d3944cR354)

TrySound replied:
Okay. Didn't know. Yeah, this is quite better. Thanks for your work.

vjeux commented:
Thanks a lot for driving this. I've probed a bunch of people and there's general agreement that this is better.

I'm currently in vacation with limited computer time but I'll get a review in the next few days.

Thanks a lot!

vjeux commented:
Could you add tests with comments in a bunch of places with all those combinations, they may need some additional treatment.

suchipi commented:
Will do

azz commented on tests/jsx/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -1,5 +1,84 @@
+</div>;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+<div>
+  {isCool ? (
```
Shouldn't this one stay on one line?

suchipi replied:
I made the decision to *always* break conditional expressions that use the JSX-style formatting. The motivation behind this was to make sure that the editor workflow described by @ryanflorence in #2208 would always be available.

I was mostly trying to avoid a situation like this:
```js
isCool ? <span>too small to break</span> : (
  <div>something that is long enough to break</div>
)
```
Which would be similar to allowing the following situation:
```js
if (isCool) "too small to break" else {
  "something that is long enough to break"
}
```
However, thinking about it again now, I could achieve the desired behavior using `group`/`ifBreak` just fine. If I made that change, this conditional expression would stay on one line (without breaking), but heterogeneous breaking (between the consequent and alternate) would be disallowed.

suchipi replied:
Update: I tried to group this properly so that the example in question would stay on one line, but ran into issues like I described (where one half of the conditional expression is multiline, and the other is not). This looked particularly wonky with chained conditional expressions. I can go into more detail, but the tl;dr is I couldn't get it working, and was starting to build more and more complex rules re: ConditionalExpression in JSX, that made it hard to predict what the code was going to do.

It may be worth revisiting this logic in the future, but in the meantime, this PR seems to be an improvement over the current implementation, so I'd like to punt on that improvement for now if possible.

azz replied:
Sounds good. Thanks for looking into that.

vjeux replied:
The way to solve this is to "linearize" the nested conditionals. You can look at printMemberChain for an example.

Once it's in a flat array, it's easier to put only a single group around, to check if there's a JSX element inside...

We'll need to get this in before the release (doesn't have to be in this PR), otherwise we're going to churn people where we change their code in one way, and then back to another.

suchipi replied:
Makes sense; I'll take a look at `printMemberChain`; sounds like exactly the pattern I need.

rattrayalex replied:
Personally, I'm not so sure that the #2208 workflow is needed for non-JSX literals. The parens around string and null literals seem quite ugly to me, and not something likely to ever be written by a human voluntarily 😄 

In particular, `{isCool ? "cool" : "very nice"}` seems far better than what we have here, and isn't uncommon in actual code. 

That said, there would be some tricky consistency issues, eg; 

```jsx
{showTheThing ? (
  <Thing />
) : null}
```
seems awkward/inconsistent, though 
```jsx
{showTheThing ? (
  <Thing />
) : (
  null
)}
```
seems bloated/verbose

suchipi replied:
In the case of a null literal as the alternate:
```jsx
{showTheThing ? (
  <Thing />
) : null}
```
I've written that by hand before. I prefer this where possible:
```jsx
{showTheThing && (
  <Thing />
)}
```
But using a ternary with a hanging null protects against creating a child with the value `undefined` (which is disallowed in React). So I can only use `&&` if I know that `showTheThing` will be false or null when not present (and never undefined).

In the case of the string literals, I've voluntarily written that syntax as a human :wink: 
But I think it would be better to make it so that the ternary only breaks and adds parens if it needs to (and one-lines it without parens otherwise). Right now we're *always* breaking and adding parens, but I want to change it to put a group around outermost ConditionalExpression (in a chain of ConditionalExpressions, possibly), and break that whole group so that it's either all on one line, or all of the consequents/alternates have parens around them.

rattrayalex commented:
This is great. I'm a little hesitant about the ternary syntax, but the edit-workflow case made in #2208  seems compelling, and there isn't really a good answer anyway. 

Sounds like @vjeux is on vacation, I may be able to review+merge.

azz commented:
> I'm a little hesitant about the ternary syntax.

Yeah it seems strange to have two styles within prettier:

JS
```js
props.showTheOtherThing
  ? "I am here"
  : "blah";
```

JSX
```jsx
{props.showTheOtherThing ? (
  <div>I am here</div>
) : (
  <div attr="blah" />
)}
```

I'm wondering if we can keep the `?:` on the LHS and still keep that workflow? Something like:

```jsx
{props.showTheOtherThing
  ? (
    <div>I am here</div>
  )
  : (
    <div attr="blah" />
  )
}
```

suchipi commented on src/printer.js:
```diff
@@ -1224,21 +1225,81 @@ function genericPrintNoParens(path, options, print, args) {
+        // Even though they don't need parens, we wrap (almost) everything in
+        // parens when using ?: within JSX, because the parens are analagous to
+        // curly braces in an if statement.
+        const wrap = node =>
```
Yeah, I'll change this

suchipi commented on src/printer.js:
```diff
@@ -1224,21 +1225,81 @@ function genericPrintNoParens(path, options, print, args) {
+        const wrap = node =>
+          concat(["(", indent(concat([hardline, node])), hardline, ")"]);
+
+        const NO_WRAP_TYPES = {
```
I personally find it easier to read, but I'm fine doing whichever style

suchipi commented on src/printer.js:
```diff
@@ -1224,21 +1225,81 @@ function genericPrintNoParens(path, options, print, args) {
         concat([
           path.call(print, "test"),
-          parent.type === "ConditionalExpression" ? printed : indent(printed)
+          parent.type === "ConditionalExpression" || forceNoIndent
```
Yeah, that would work

suchipi commented:
The problem with the proposed ternary style:
```jsx
{props.showTheOtherThing
  ? (
    <div>I am here</div>
  )
  : (
    <div attr="blah" />
  )
}
```
is that I've never seen a human write that.

I don't think it's too much of an issue to have two ternary styles, because ternaries are used in a pretty different way in JSX than they are elsewhere; they're standing-in for If/Else.
The equivalent If/Else syntax to the proposed ternary syntax would be:
```jsx
if (foo) {
  return "foo";
}
else {
  return "bar";
}
```
Which @jlongster prefers iirc but the decision was made to not do that because the rest of the JS community is doing `} else {`.

So I think in this case, sticking with the syntax the community is already using would be best. They're both ConditionalExpression nodes... but people think about them differently.

That said, I recognize that having both syntaxes could look weird if they're right next to each other:
```jsx
const a = b
  ? "c - really really long, long enough to break this ternary"
  : "d"; // normal mode

const e = f ? ( // JSX mode, because the ternary contains at least one JSXElement
  <g />
) : (
  <h />
);
```

We could look into only enabling JSX-mode conditional expressions if they're inside a JSXExpressionContainer, so that that wouldn't happen, but I'm enabling JSX-mode outside of JSXExpressionContainers right now to facilitate this pattern which I have seen before:

```jsx
render() {
  const contentLeft = childrenLeft ? (
    <div className="left-wrapper">
      {childrenLeft}
    </div>
  ) : (
    <div className="left-wrapper no-items">
      No Left Items
    </div>
  );

  const contentRight = childrenRight ? (
    <div className="right-wrapper">
      {childrenRight}
    </div>
  ) : (
    <div className="right-wrapper no-items">
      No Right Items
    </div>
  );

  return (
    <div className="some-wrapper">
      {contentLeft}
      {contentRight}
    </div>
  );
}
```
Namely, assembling JSX "fragments" that will be put into "slots" in the final render output. In this case, *most* of the JSX is not within a JSXExpressionContainer.

You can technically use if/else in this context, but then you need to use `let` (or var) instead of `const`, which is why I think ternaries are preferred (also because you can move them down into the render output directly without rewriting them).

suchipi commented:
I have added a commit that makes the following changes:
* JSX-mode ternaries will now fit on one line if they can
* JSX-mode ternaries will now only group around the outermost ternary, so if anything breaks, the whole ternary chain breaks and parens are added
* `null` is not wrapped with parens in JSX-mode ternaries, even when breaking.

Please see [this file](https://github.com/suchipi/prettier/blob/60138bf517c9abebacfda25a636fcb118a016e9b/tests/jsx/conditional-expression.js) for a break-down.

vjeux commented on tests/arrows/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -304,10 +304,11 @@ export const bem = block =>
       {info.item.widget.missingProp}
-    </span>}
+    </span>
+  )}
```
This is great, people complained about the fact that `}` was on the same line as `}` as well.

rattrayalex commented:
This is really interesting. I don't think it'll be universally popular and I get the feeling it'll receive some iteration, but I'm in pretty strong support of giving this a try.

suchipi commented:
Sounds like the general consensus is in favor of these changes- is there anything else that should be addressed, or is this ready to merge? (Not ruling out future iteration with the help of the community)

VladShcherbin commented on tests/jsx/logical-expression.js:
```diff
@@ -0,0 +1,23 @@
+      <div></div>
+    </div>
+  </span>}
+</div>;
```
I'm curious about the last one with multiline. Should it be like this?

```jsx
{isVisible && <p>is visible</p>}

{isVisible && (
  <div>
    <p>is visible</p>
  </div>
)}

VladShcherbin commented:
All examples look awesome, exactly the style how a lot of devs write jsx. Can't wait it to land in prettier together with https://github.com/prettier/prettier/pull/2442.

levithomason commented:
Indeed, once this and #2442 are landed I'm going to give this a run over at [Semantic-UI-React](https://github.com/Semantic-Org/Semantic-UI-React) as well.

rattrayalex commented:
@suchipi apologies for the delay - I'm not aware of blockers, would you mind rebasing and updating the snapshots to fix conflicts? Then ping me and I'll merge 😄

suchipi commented:
@rattrayalex rebased and snapshots updated

jlongster commented:
fwiw this change looks good to me, thanks for reviewing the code @rattrayalex. I'd like to look at nested ternaries in the general JS case anyway later.

rattrayalex commented:
Awesome!

suchipi commented:
Thanks for the community involvement on this one, everyone! Looking forward to seeing it released.

kentcdodds commented:
What can we do to help get this released? I'm anticipating this release more than any previous release. I'm very excited about this :smile:

VladShcherbin commented:
@kentcdodds it was already merged, we just need to wait for https://github.com/prettier/prettier/issues/2589 and 1.6 to be released I guess.

kentcdodds commented:
I missed that issue. Will track that. Thank you :smile:

dodekeract commented:
Why is this mandatory? I agree that we should respect the developer's choice here, but why do we disrespect it if the dev didn't insert parentheses?

suchipi commented:
In order to format your code, prettier parses your code into an AST, then removes everything in the file and reprints it by hand. Unfortunately, parens are not included in the AST, because they don't have semantic significance to the language (as far as the engine is concerned; ASTs are traditionally created so that the engine can understand what you wrote). But that means that when we go through the AST to reprint everything, we have to decide to put parens or not, since that original information isn't there.

We could add in some hacks to look at what the code was like before we parsed it, and then decide to put parens in based on that (which is what we did for deciding if we should print an object on one line or multiple), but that's generally against the philosophy of prettier, because we are trying to normalize format as much as possible; the goal of this tool is to remove wiggle room so that teams who want a consistent format style can get one and stop arguing about stylistic preferences.

Parens kinda don't get as much attention as they deserve because they're harder to work with due to how the AST treats them, which is unfortunate. I think that if parens were in the AST, we would try to preserve what the user wrote; it's an arbitrary distinction.

Are there specific cases where parens are being inserted that you dislike? Those might be bugs that we need to fix.

dodekeract commented:
@suchipi Okay, that's a fair point for not respecting every *local* choice. Are there any reasons against keeping this as a *global* option? I think arrow functions with unnecessary `()` are rather ugly.

suchipi commented:
Although we have already added a few, we are generally hesitant to add global options, because the more options that we have to support, the harder the codebase is to maintain, because the complexity of the problem space increases, even if we are only adding boolean options:

If we only had one global option, let's just call it "A" for this thought exercise, then we would need to make sure that code formats correctly for these cases:
  * A on
  * A off

Adding a new option called "B" would mean that we would need to make sure all these combinations format correctly:
  * A on, B on
  * A off, B on
  * A on, B off
  * A off, B off

For three options (adding "C"), it becomes:
  * A on, B on, C on
  * A on, B on, C off
  * A on, B off, C on
  * A on, B off, C off
  * A off, B on, C on
  * A off, B on, C off
  * A off, B off, C on
  * A off, B off, C off

So you can see that each new option increases the complexity of the codebase.

We can automate testing around all these different option combinations, but it's really easy for strange edge cases to slip through, because there are so many different ways to write code; there are a lot of patterns we aren't familiar with, so we might break things even if all our tests are passing.

So in general, we are hesitant to add more options. If most of the community is comfortable with one style, we would prefer to not add an option, and instead support the majority. However, there are a few things that the community is very divided about (tabs versus spaces, semicolons versus no semicolons, etc), for which we have added options.

yuritoledo commented:
+ 1

![image](https://user-images.githubusercontent.com/18474538/41536043-c3a61fe4-72da-11e8-931a-5976e36a8282.png)

duailibe commented:
@yuritoledo Please open a new issue, filling the template with the necessary info

yuritoledo commented:
the #4714 was mine ;)
